logr.go

```diff
@@ -429,6 +429,15 @@ type LogSink interface {
 	WithName(name string) LogSink
 }
 
+// FlushLogSink is an optional interface that LogSinks can implement when they
+// support buffering of log messages.
+type FlushLogSink interface {
+	// Flush should be called by a program shortly before terminating and,
```

Comment:
The problem I have with this is that it offloads cognitive burden onto the end-user.

If I was reading the docs for logr for the first time, I would see this and think: "I need to pick my implementation, create it, then test if it supports Flush, and if so set up a goroutine with a timer, plus ensure that I never call Exit(), plus I should catch panic(), or my logs will get lost."

That's a TON of burden for someone who just wants to write some logs.

IMO the saner approach is to:
* impl docs/API gives users a choice to buffer (better perf, more complex) or not (simpler)
* IFF they choose to buffer:
  - set up a goroutine internally
  - return a `func()` that the user should try to call when they exit

This puts  the burden of calling the func on those users who choose to use a buffered impl.   Other users will not have to think about it.

So something like:

```
log, flush := impl.NewBuffered(flushPeriod)
defer flush()
```

Reply:
We can't do it like that in Kubernetes. But okay, let's keep this out of logr itself.

Reply:
Does this mean you also don't want this in zapr? See https://github.com/go-logr/zapr/pull/47

That then will be a problem in Kubernetes, because `component-base/logs` can only store a `logr.Logger` in klog, but not a logger plus a custom flush function.

Reply:
> We can't do it like that in Kubernetes. But okay, let's keep this out of logr itself.

I'd like to see how bad the lack of this API is for real consumers like k8s.  If you can show me that it's terrible, and that doing this (or something like it) would be better, I am open to discuss.

Reply:
I'm not against this in zapr - that's on the other side of the "do I want buffering" decision.  I still think impls should encapsulate it as much as possible.  We control klog - why can't it keep a flush function?

Reply:
It's doable, but I think the API would be odd. The [`SetLogger`](https://github.com/kubernetes/klog/blob/331d2323a192188a9aca7006c8dedd0c61c03833/klog.go#L859-L873) function is part of the stable klog v2 API, we can't change it.

We could add `SetLoggerWithFlushing` or something like that, but that feels sub-optimal when we can get the same effect by defining an interface for the LogSink somewhere.

If we want to do it entirely without such an interface, then I probably would leave klog unchanged and handle it entirely in `component-base/logs`: it can use `klog.SetLogger` for logging and then keep a separate function pointer inside the package for the flush function.

Reply:
I think that sounds right.  If a user did it inside `main()`, they would assume responsibility for calling the final flush, right?  component-base is effectively `realMain()`